reVX/offshore/dist_to_ports.py

```diff
@@ -397,21 +431,227 @@ def run(cls, ports, excl_fpath, input_dist_layer='dist_to_coast',
                     .format(ports))
         dtp = cls(ports, excl_fpath, input_dist_layer=input_dist_layer)
-        if output_dist_layer is None:
-            output_dist_layer = os.path.basename(ports).split('.')[0]
 
         dist_to_ports = None
         if update_layer:
             hsds = check_res_file(excl_fpath)[1]
             with ExclusionLayers(excl_fpath, hsds=hsds) as tif:
-                if output_dist_layer in tif.layers:
+                if output_dist_layer in tif:
                     dist_to_ports = tif[output_dist_layer]
 
         if dist_to_ports is not None:
             logger.info('Updating exising layer {}'.format(output_dist_layer))
 
         dist_to_ports = dtp.least_cost_distance(dist_to_ports=dist_to_ports,
                                                 max_workers=max_workers)
-        dtp.save_as_layer(output_dist_layer, dist_to_ports, chunks=chunks)
+
+        if output_dist_layer:
+            dtp.save_as_layer(output_dist_layer, dist_to_ports, chunks=chunks)
 
         return dist_to_ports
+
+
+class AssemblyAreas:
+    """
+    Class to compute the distance from port to assembly areas using the
+    distance to port arrays produced using DistanceToPorts
+    """
+    DIST_COL = 'dist_p_to_a'
+
+    def __init__(self, assembly_areas, excl_fpath):
+        """
+        Parameters
+        ----------
+        assembly_areas : str | pandas.DataFrame
+            DataFrame or path to csv or json containing assembly area
+            meta and locational data
+        excl_fpath : str
+            Path to exclusions .h5 file with distance to coast layer
+        """
+        self._assembly_areas = parse_table(assembly_areas)
+        self._excl_fpath = excl_fpath
+        self._assembly_idx = self._get_assembly_array_idx(self._assembly_areas,
+                                                          excl_fpath)
+
+    def __repr__(self):
+        msg = ("{} with {} areas"
+               .format(self.__class__.__name__, len(self.assembly_areas)))
+
+        return msg
+
+    @property
+    def assembly_areas(self):
+        """
+        DataFrame with assembly area meta and locational data
+
+        Returns
+        -------
+        pandas.DataFrame
+        """
+        return self._assembly_areas
+
+    @staticmethod
+    def _build_tree(excl_fpath):
+        """
+        Build cKDTree from exclusions coordinates
+
+        Parameters
+        ----------
+        excl_fpath : str
+            Path to exclusions .h5 file with distance to coast layer
+
+        Returns
+        -------
+        tree : cKDTree
+            cKDTree build on exclusions coordinates
+        shape : tuple
+            Shape of exclusion layers/arrays
+        """
+        with ExclusionLayers(excl_fpath) as f:
+            lat = f['latitude']
+            lon = f['longitude']
+
+        lat_lon = np.dstack((lat.ravel(), lon.ravel()))[0]
+
+        # pylint: disable=not-callable
+        tree = cKDTree(lat_lon)
```

Comment:
Is this slow? Could we only make a tree on the offshore exclusion points?

Reply:
Fixed, not to bad, great minds!